### PR TITLE
WDP-90

### PR DESCRIPTION
**Pull request by lukaszcwiklinski**
https://kindergarten.stxnext.pl/browse/WDP-90
https://gist.github.com/lukaszcwiklinski/fa0ca1b0eb316730577905d206baa45c

### CODE REVIEW

File: src/presence_analyzer/utils.py [hunk 1]
```diff
@@ -105,6 +107,73 @@ def get_data():
...
+                user_id = int(row[0])
+                date = datetime.strptime(row[1], '%Y-%m-%d').date()
+                start = datetime.strptime(row[2], '%H:%M:%S').time()
+                end = datetime.strptime(row[3], '%H:%M:%S').time()
+            except (ValueError, TypeError):
+                log.debug('Problem with line %d: ', i, exc_info=True)
+
+            data.setdefault(date, {})[user_id] = interval(start, end)
+
+    return data
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: I don't like this approach at all. Read data once. Cache it and then on each end point filter it using ur helper functions.

File: src/presence_analyzer/utils.py [hunk 2]
```diff
@@ -105,6 +107,75 @@ def get_data():
...
+    Converts representation of data return from get_data().
+
+    Returns: Dict of dicts eg:
+    data = {
+        datetime.date(2013, 10, 1): {
+            10: 31395,
+            12: 45863
+        }
+    }
+    """
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: Returns: Dict of dicts eg:

    {
        datetime.date(2013, 10, 1): {
             10: 31395,
             12: 45863
        },
        (...)
    }

File: src/presence_analyzer/utils.py [hunk 3]
```diff
@@ -158,8 +229,27 @@ def group_by_weekday(items):
...
+    month_total = {}
+    for date, values in items:
+        for user, worktime in values.iteritems():
+            if user not in month_total:
+                month_total[user] = 0
+
+            month_total[user] += worktime
+
+    result = dict(Counter(month_total).most_common(5))
+    sorted_result = sorted(result.items(), key=lambda x: x[1], reverse=True)
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: return it directly without

File: src/presence_analyzer/views.py [hunk 4]
```diff
@@ -92,6 +98,46 @@ def presence_weekday_view(user_id):
...
+    Return list of lists, eg:
+    [
+        [2011, 7, "2011 - July"],
+        [2012, 9, "2012 - September"],
+        ...
+    ]
+    """
+    data = get_year_and_months()
+    result = [
+        (item['year'], item['month'], item['date'])
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: why not just item.values() and wrap it wit `sorted`

File: src/presence_analyzer/views.py [hunk 5]
```diff
@@ -92,6 +98,46 @@ def presence_weekday_view(user_id):
...
+        for item in data
+    ]
+    sorted_result = sorted(result)
+    unique_result = [list(x) for x in set(tuple(x) for x in sorted_result)]
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: sth more descriptive then x and use map

File: src/presence_analyzer/templates/top_five_by_month.html [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                            ];
+                        });
+                        chartValues.unshift([
+                            {type: 'string', label: 'User'},
+                            {type: 'number', label: 'Time in (h)', p: {html: true}},
+                            {type: 'string', role: 'style'},
+                            {label: 'Avatar', role: 'tooltip', p: {html: true}}
+                        ]);
+
+                         var data = google.visualization.arrayToDataTable(chartValues),
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/top_five_by_month.html]: vars at the top of the scope...

File: src/presence_analyzer/templates/top_five_by_month.html [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                            colors = ['#b9f2ff', '#e5e4e2', '#ffd700', '#c0c0c0', '#cd7f32'],
+                            options = {
+                                tooltip: {isHtml: true}
+                            };
+
+                        $.each(result, function (index, item) {
+                            var name = nameAndAvatarById[item[0]][0],
+                                time = formatSeconds(item[1]),
+                                color = colors[index],
+                                avatar = 'Time: ' + formatSeconds(item[1]) + '<br /><img src="'+nameAndAvatarById[item[0]][1]+'">';
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/top_five_by_month.html]: Still wrong. Make sth like 
```
avatar = [
    'Time: ',
    formatSeconds(time),
    '<br />'
    '<img src=" + src + ">'
].join ('');
```
U have declared time and not using it, why

File: src/presence_analyzer/templates/top_five_by_month.html [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+                $dropdown.show();
+                $loading.hide();
+            });
+
+            $('#user-id').change(function () {
+                var selectedUser = $('#user-id').val(),
+                    $chartDiv = $('#chart-div'),
+                    $avatar = $('#avatar'),
+                    date = selectedUser.split('-');
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/top_five_by_month.html]: `selectedUser.replace('-', '/')` and directly to url?

File: src/presence_analyzer/templates/top_five_by_month.html [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                if (selectedUser) {
+                    $loading.show();
+                    $chartDiv.hide();
+                    $avatar.hide();
+
+                    $.getJSON('/api/v1/top_five/' + date[0] + '/' + date[1], function (result) {
+                        var chart,
+                            chartValues = [],
+                            colors = ['#b9f2ff', '#e5e4e2', '#ffd700', '#c0c0c0', '#cd7f32'],
+                            data,
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/top_five_by_month.html]: move afterr `chart`

File: src/presence_analyzer/views.py [hunk 10]
```diff
@@ -92,6 +96,45 @@ def presence_weekday_view(user_id):
...
+    Return list of dicts, eg:
+    [
+         ['2011 - December', 12, 2011],
+         ['2011 - July', 7, 2011],
+         [...],
+    ]
+    """
+    data = get_year_and_months()
+    unique_result = []
+    result = [
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: You don't need assign it to a variable sinc ur not using it

File: src/presence_analyzer/views.py [hunk 11]
```diff
@@ -92,6 +96,46 @@ def presence_weekday_view(user_id):
...
+@app.route('/api/v1/top_five/<int:year>/<int:month>', methods=['GET'])
+@jsonify
+def top_five_worktime(year, month):
+    """
+    Returns top 5 work time for users grouped by date.
+    """
+    worktime = get_data_by_date()
+    result = [
+        (date, values)
+        for date, values in sorted(worktime.iteritems())
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: `sorted(worktime.iteritems())` sorted wont work on generetor so is necessary at all?

File: src/presence_analyzer/views.py [hunk 12]
```diff
@@ -92,6 +96,46 @@ def presence_weekday_view(user_id):
...
+         ['2011 - July', 7, 2011],
+         [...],
+    ]
+    """
+    data = get_year_and_months()
+    unique_result = []
+
+    [
+        unique_result.append(item.values())
+        for item in sorted(data) if item.values() not in unique_result
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: mb move sroted to  `sorted(get_year_and_months())`

File: src/presence_analyzer/utils.py [hunk 13]
```diff
@@ -153,13 +230,31 @@ def group_by_weekday(items):
...
+    for date, values in items:
+        for user, worktime in values.iteritems():
+            if user not in month_total:
+                month_total[user] = 0
+
+            month_total[user] += worktime
+
+    result = dict(Counter(month_total).most_common(5))
+
+    return sorted(result.items(), key=lambda x: x[1], reverse=True)
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: use `itemgetter`

File: src/presence_analyzer/utils.py [hunk 14]
```diff
@@ -105,6 +112,76 @@ def get_data():
...
+
+    return data
+
+
+def get_data_by_date():
+    """
+    Converts representation of data return from get_data().
+
+    Returns: Dict of dicts eg:
+    data = {
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: remove `data = `

File: src/presence_analyzer/templates/nav.html [hunk 15]
```diff
@@ -3,18 +3,23 @@
     <ul>
         <li
         % if request.path == '/presence_weekday' or request.path == '/':
```
Comment (dpolcyn) [inline src/presence_analyzer/templates/nav.html]: refactor it, it should be loop
if you gona have 20 tabs then u will put 20 if's?

File: src/presence_analyzer/tests.py [hunk 16]
```diff
@@ -8,12 +8,11 @@
 import httplib
 import json
 import os.path
-import time
 import unittest
 
-import main
-import utils
-import views
+import presence_analyzer.main as main
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: from presence_analyzer import main?
i dont know why you use 'as' here if you dont import similar name module or long name module

File: src/presence_analyzer/utils.py [hunk 17]
```diff
@@ -153,13 +231,31 @@ def group_by_weekday(items):
...
         start = items[date]['start']
         end = items[date]['end']
         result[date.weekday()].append(interval(start, end))
+
     return result
 
+
+def top_five(items):
+    """
+    Summs month top 5 worktime for user.
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: use defaultdict

File: src/presence_analyzer/views.py [hunk 18]
```diff
@@ -92,6 +99,46 @@ def presence_weekday_view(user_id):
     result.insert(0, ('Weekday', 'Presence (s)'))
     return result
 
+
+@app.route('/api/v1/years_and_months/', methods=['GET'])
+@jsonify
+def presence_years_and_months():
+    """
+    Returns work time for users grouped by date.
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: by month

File: src/presence_analyzer/templates/top_five_by_month.html [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+
+                $dropdown.show();
+                $loading.hide();
+            });
+
+            $('#user-id').change(function () {
+                var selectedUser = $('#user-id').val(),
+                    $chartDiv = $('#chart-div'),
+                    $avatar = $('#avatar'),
+                    date = selectedUser.replace('-', '/');
```
Comment (dpolcyn) [inline src/presence_analyzer/templates/top_five_by_month.html]: why ?

File: src/presence_analyzer/tests.py [hunk 20]
```diff
@@ -341,6 +388,68 @@ def test_get_xml_data(self):
...
+            data[0], ['date', 'month', 'year']
+        )
+
+    def test_get_data_by_date(self):
+        """
+        Test wokrtime for user_id in date.
+        """
+        data = utils.get_data_by_date()
+
+        self.assertIsInstance(data, dict)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: you are checking it below

File: src/presence_analyzer/templates/top_five_by_month.html (hunk 19 above)
Comment (lukaszcwiklinski) [inline src/presence_analyzer/templates/top_five_by_month.html]: to pass date to url in format 2011/8

File: src/presence_analyzer/tests.py [hunk 21]
```diff
@@ -180,16 +176,58 @@ def test_presence_by_start_end(self):
...
+        self.assertEqual(api.status_code, httplib.OK)
+        self.assertEqual(result, data)
+
+    def test_top_five_view(self):
+        """
+        Test result of available top five work times
+        in specific year and month.
+        """
+        api = self.client.get(
+            '/api/v1/top_five/2011/8',
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: 0 instead of 8

File: src/presence_analyzer/utils.py [hunk 22]
```diff
@@ -105,6 +113,76 @@ def get_data():
...
+def get_year_and_months():
+    """
+    Extracts presence data from CSV file and takes unique years and months.
+
+    Returns: list of dicts eg:
+    [
+        {
+            'year': 2011,
+            'month': 8,
+            'date': "2011 - August"
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: "???????????????????????

**Pull request merged**